### ISSUE TITLE
Improvement to submission overload risk

### ISSUE THREAD

**Issue by aikar** [Bastian/bStats#132]
Regarding https://github.com/PaperMC/Paper/pull/4914

While this was a good change on our end, I have a suggestion for improving it on your end.

The library p-queue is my favorite promise based queue library for NPM, highly used.

Use this library to throttle submission tasks to ensure only X amount can be concurrently processing at a time.

This should be a rather simple and quick change (Sorry don't have time to PR this, but I'm sure you get the gist of what I'm suggesting)

Furthermore, bstats submissions are not end of the world if the request is dropped.

I suggest once the submission has been received over the http endpoints, to not await the promise for completion of the submission, and just quickly tell the submitter "got it" once validation is done, submitting the data into the p-queue to fire when ever its turn comes.

Shutdown of the process does need to block until p-queue's instance .onIdle() method resolves to ensure all pending tasks are flushed before shutdown finishes.

Finally, one final optimization that could be made is if the p-queue size > X, to write the request data to a temporary data file on disk, unload the request object from memory (releasing references to the object), and submit into the p-queue a task that will load the data from FS before proceeding.

You could just do this for all requests to keep it simple, then the p-queue tasks only hold a reference to a temporary file name, and shutdown could let those pending files stay on disk and then only need to flush in-processing tasks at shutdown.

This would guarantee memory stays light even if 10k pending submissions are waiting.

Happy to provide a code review if you need for this (I've done this kind of logic for my own work use plenty)

**Comment by Bastian** [comment]
Thank you for the suggestion! 🚀 

`p-queue` indeed looks like an interesting library, but I'm unsure if there are many benefits of queueing data submissions.
NodeJS being single-threaded is already a big help at CPU-restricting the application. On my 16-core server, I have 12 instances of the bStats backend running. This means that there are always 4 free cores available for other workloads (i.e., the database and frontend).

This means, that a queue would only help to free up load for other tasks of the backend itself (e.g., serving data for the charts on the website). However, I feel like there are easier approaches to achieve the same result, e.g. by using nginx to route data submission requests to another backend instance than all the other requests.

The idea to temporarily store requests in files is interesting, but at the moment the problem is not that big and with the changes to the Metrics classes, it should completely dissolve in the long term (even though it might take a few months/years). 

PS: 
> Sorry don't have time to PR this

It's good that you didn't PR to this repo. I'm currently rewriting the whole bStats backend and frontend and the data submission is already done by the [new backend](https://github.com/Bastian/bstats-backend) and no longer by the code in this repository. I feel like I should probably add a warning in the README (and maybe PR template) to ensure that noone PRs to the outdated code.

**Comment by aikar** [comment]
I took it from your Paper PR that you were having load spike issues from lots of people submitting requests at the same time. The queue would basically achieve the same thing as clients staggering uploads by forcing it on the server side, and nodejs manages idle waiting connections pretty cheaply.

If not, then what was the trigger of that investigation & PR?

Though even if load is usually fine, it can be a good defensive technique to ensure a large bomb of requests doesn't overload and trigger instability.

**Comment by Bastian** [comment]
> If not, then what was the trigger of that investigation & PR?

It was more of a "fix it before it becomes a problem".
There is a spike in the CPU usage at xx:05 / xx:35 (and at xx:00 / xx:30 but for another reason) but nothing that causes a problem **yet**.
![image](https://user-images.githubusercontent.com/5033001/102920406-05992e80-448b-11eb-9d4a-bcc7537f1f42.png)


> Though even if load is usually fine, it can be a good defensive technique to ensure a large bomb of requests doesn't overload and trigger instability.

I agree but under normal conditions (with the new Metrics classes), the load should always be even for data submission. There would only be a problem if someone intentionally sends tons of requests (i.e., a DDoS attack) but in that case there are probably better (non-ratelimited) endpoints to target than the ones for data submission.

At the moment, I just don't feel the need for a queue and all the possible problems that come with it. E.g., what happens with data that was received at xx:29 and processed at xx:31? Yes, I could store the submission timestamp but than I run into another problem: The chart data might not be up-to-date immediately at xx:00 / xx:30 on the website and I have to delay it for a few seconds or even minutes.

Nothing that cannot be solved but I prefer to keep it simple for now and just process the requests as they come in. If the changes to the Metrics classes turn out to not reduce the CPU spike, I will reconsider the queue.

> I suggest once the submission has been received over the http endpoints, to not await the promise for completion of the submission, and just quickly tell the submitter "got it" once validation is done, submitting the data into the p-queue to fire when ever its turn comes.

To come back to this suggestion: I think this is a good idea. Checking ratelimits and then simply returning a `202 Accepted` is probably enough.
